python/grass/grassdb/history.py

```diff
@@ -9,113 +9,359 @@
-    if not history_path:
-        history_path = get_current_mapset_gui_history_path()
     try:
-        if os.path.exists(history_path):
-            file_history = open(history_path, encoding="utf-8", mode="a")
-        else:
-            file_history = open(history_path, encoding="utf-8", mode="w")
-        file_history.write(command + "\n")
+        with open(history_path, encoding="utf-8", mode="r+") as file_history:
+            try:
+                history_entries = json.load(file_history)
+                # Check if the index is valid
+                if 0 <= index < len(history_entries):
+                    # Remove the entry at the specified index
+                    del history_entries[index]
+                # Write the modified history back to the file
+                file_history.seek(0)
+                file_history.truncate()
+                json.dump(history_entries, file_history, indent=2)
+            except ValueError as ve:
+                raise ValueError(
+                    _("Error decoding content of JSON history file {}").format(
+                        history_path
+                    )
+                ) from ve
     except OSError as e:
         raise OSError(
-            _("Unable to add entry to history file {}").format(history_path)
+            _("Unable to remove entry from JSON history file {}").format(history_path)
         ) from e
-    finally:
-        if file_history:
-            file_history.close()
 
 
-def remove_entry_from_history(del_line_number, history_path=None):
+def remove_entry(history_path, index):
     """Remove entry from history file.
 
-    :param int del_line_number: line number of the command to be removed
-    :param str|None history_path: history file path string
+    :param str history_path: path to the history log file
+    :param int index: index of the command to be removed
+    """
+    if get_history_file_extension(history_path) == ".json":
+        _remove_entry_from_JSON(history_path, index)
+    else:
+        _remove_entry_from_plain_text(history_path, index)
+
+
+def convert_plain_text_to_JSON(history_path):
+    """Convert plain text history log to JSON format.
+
+    :param str history_path: path to the history log file
     """
-    file_history = None
-    if not history_path:
-        history_path = get_current_mapset_gui_history_path()
+    if get_history_file_extension(history_path) != ".json":
+        try:
+            lines = _read_from_plain_text(history_path)
+
+            # Extract file path and name without extension
+            file_path = history_path.parent
+            file_name_no_ext = history_path.stem
+
+            # JSON file path
+            json_path = file_path / (file_name_no_ext + ".json")
+
+            # Write JSON data to a new file
+            with open(json_path, "w") as json_outfile:
+                json.dump(lines, json_outfile, indent=2)
+
+            # Remove the old plain text file
+            history_path.unlink()
```

Comment:
How about keeping the old file there? Is there any reason not to? When running in a mapset with an older version, it could still be using the plain text one. In that case you would probably have to change `get_current_mapset_gui_history_path` to primarily return json, not plain text file.